File: DSA/Assignment_6/stack_arr.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Stack
{
    int top;
    int cap;
    int *arr;
}Stack;

Stack *createStack(int cap)
{
    Stack *stack=(Stack *)malloc(sizeof(Stack));
    stack->cap=cap;
    stack->top=-1;
    stack->arr=(int *)malloc(cap*sizeof(int));
    return stack;
}

int isFull(Stack *stack)
{
    return stack->top==stack->cap-1;
}

int isEmpty(Stack *stack)
{
    return stack->top==-1;
}

void push(Stack *stack, int ele)
{
    if(isFull(stack))
    {
        printf("Overflow!\n");
        return;
    }
    stack->arr[++stack->top]=ele;
}

int pop(Stack *stack)
{
    if(isEmpty(stack))
    {
        printf("Underflow!\n");
        return -1;
    }
    return stack->arr[stack->top --];
}

int peek(Stack *stack)
{
    if(isEmpty(stack))
    {
        printf("Stack is empty!\n");
        return -1;
    }
    return stack->arr[stack->top];
}

void print_stack(Stack *stack)
{
    if(isEmpty(stack))
    {
        printf("Stack is empty!\n");
        return;
    }
    for(int i=stack->top;i>=0;i--)
    printf("%d\n",stack->arr[i]);
}

int main()
{
    int n, ele;
    printf("Enter size of stack:");
    scanf("%d",&n);
    Stack *stack=createStack(n);
    for(int i=0;i<n;i++)
    {
        printf("Enter elements of stack:");
        scanf("%d",&ele);
        push(stack, ele);
    }
    printf("Original Stack:");
    print_stack(stack);

    //pop top element of stack
    
    return 0;
}

File: DSA/Assignment_6/stack_arr.py
class Stack:
    def __init__(self, cap):
        self.arr=[]
        self.cap=cap
        self.top=-1

    @staticmethod
    def createStack(capacity):
        return Stack(capacity)
    
    def isFull(self):
        return self.top==self.cap-1
    
    def isEmpty(self):
        return self.top==-1
    
    def push(self, ele):
        if self.isFull():
            print("Stack Overflow!")
            return
        self.arr.append=ele
        self.top+=1

    def pop(self):
        if self.isEmpty():
            print("Stack Underflow!")
            return
        return self.arr.pop()
    
    def peek(self):
        if self.isEmpty():
            print("Stack is Empty!")
            return
        return self.arr[self.top]

    def print_stack(self):
        if self.isEmpty():
            print("Stack is Empty!")
            return
        for i in range(self.top,1,-1):
            print(self.arr[i],end=' ')

def main():
    #Formation of Stack
    n=int(input("Enter size  of stack:"))
    stack=Stack.createStack(n)
    for _ in range(n):
        ele=int(input("Enter element:"))
        stack.push(ele)
    print("Original Stack:",end='')
    stack.print_stack()

    #pop top element of stack
    stack.pop()